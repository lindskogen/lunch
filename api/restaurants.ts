import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../lib/lunch-fetcher/fetch-restaurants";

export default async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};
