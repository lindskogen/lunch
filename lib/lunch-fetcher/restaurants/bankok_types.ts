export interface Ingestion {
  price: number;
  type: string;
}

export interface Item {
  id: string;
  name: string;
  preview: string;
  info: string;
  description: string;
  outOfStock: boolean;
  sortOrder: number;
  category: number;
  additionalItems: any[];
  ingestions: Ingestion[];
  itemCategory: string;
}

export interface ItemCategory {
  id: string;
  name: string;
  preview?: any;
  description: string;
  category: number;
}

export interface RestaurantData {
  items: Item[];
  itemCategories: ItemCategory[];
}
