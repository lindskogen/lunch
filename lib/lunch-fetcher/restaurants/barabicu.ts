import * as _ from "lodash";
import { isNotNull } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay, RestaurantDayMenu } from "../types";

export const url = "https://barabicu.se/";

const weekdayTranslations: WeekDay[] = [
  "Måndag",
  "Tisdag",
  "Onsdag",
  "Torsdag",
  "Fredag"
];
export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#todays-lunch li")
    .toArray()
    .map<RestaurantDayMenu>((dayNode, index) => {
      return {
        items: $("h3", dayNode)
          .toArray()
          .map<FoodItem | null>(elem => {
            if ($(elem).text() === '' || $(elem.next).text() === '') {
              return null;
            }
            return {
              name: $(elem.next).text(),
              title: $(elem).text().split('•')[0].trim()
            };
          }).filter(isNotNull),
        wday: weekdayTranslations[index]
      };
    });

  return { name: "Barabicu", url, days: menus };
};
