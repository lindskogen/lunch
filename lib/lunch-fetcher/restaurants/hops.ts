import * as _ from "lodash";
import { id, isNotNull, isNotZero, isWeekday } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay } from "../types";

export const url = "http://www.hopsbar.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#Veckansmeny")
    .toArray()
    .map(item => {
      const allNodes = _.flatMap(item.childNodes, node =>
        $(node)
          .text()
          .trim()
          .split(/\n/)
      )
        .map(s => s.trim())
        .filter(id);

      const weekMealsStartIndex = allNodes.findIndex(
        text => text === "Veckans rätter"
      );

      const weekItems: FoodItem[] = _.takeWhile(
        allNodes.slice(weekMealsStartIndex + 1),
        text => !_.includes(text, " kaffe ")
      ).map(food => {
        const foodname = food.trim();
        const matches = foodname.split(':');

        if (matches.length > 1) {
          const [title, name] = matches;
          return { name: name.trim(), title: title.trim() };
        }

        return { name: foodname };
      });

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const items = foods.map(food => ({ name: food.trim() }));

          return {
            wday: wday as WeekDay,
            items: _.takeWhile(
              items,
              ({ name }) => name !== "Veckans rätter"
            ).concat(weekItems)
          };
        });
    });

  return { name: "Hops", url, days: _.flatten(menus) };
};
