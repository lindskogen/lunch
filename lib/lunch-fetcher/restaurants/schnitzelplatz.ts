import {take, drop} from "lodash-es";
import { Restaurant, FoodItem, RestaurantDayMenu, WeekDay } from "../types";
import { CheerioAPI } from "cheerio";

export const url = "https://schnitzelplatz.se/lunch/";

const firstPartOfSplitOrAll = (str: string, split: string): string => {
  const parts = str.split(split);
  if (parts && parts.length > 1) {
    return parts[0];
  } else {
    return str;
  }
};

export const parseHtml = ($: CheerioAPI): Restaurant => {
  const menus = $(".introbox__container h4")
    .get()
    .map(node => [
      $(node).text().trim(),
      firstPartOfSplitOrAll(
        $(node)
          .next("p")
          .text(),
        "\n"
      )
    ]);

  const numWeekItems = menus.length - 5;

  const weekItems: FoodItem[] = take(menus, numWeekItems).map(
    ([title, name]) => ({
      title: title.toUpperCase(),
      name
    })
  );
  const dayItems = drop(menus, numWeekItems);

  const days: RestaurantDayMenu[] = dayItems.map(([day, name]) => ({
    wday: day as WeekDay,
    items: [{ name }].concat(weekItems)
  }));

  return { name: "Schnitzelplatz Lagerhuset", url, days };
};
