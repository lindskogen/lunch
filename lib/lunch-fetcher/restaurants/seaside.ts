import * as _ from "lodash";
import { isWeekday, isNotNull, isNotZero } from "../lib/utils";

export const url = "http://seaside.kvartersmenyn.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $(".day .meny")
    .toArray()
    .map(item => {
      const allNodes = item.childNodes
        .filter(i => i.tagName === "strong" || i.nodeValue)
        .map(item => $(item).text());

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const foodsRegex = /^([A-Ã–]+( [0-9])?) ?[\n :]([^0-9:\n]+)/gm;

          const allFoodsForWeekDay = foods.join("\n");

          const parsedFoods = [];

          let res = foodsRegex.exec(allFoodsForWeekDay);
          while (res !== null) {
            const [, title, , name] = res;
            parsedFoods.push({ title, name: name.trim() });
            res = foodsRegex.exec(allFoodsForWeekDay);
          }

          return {
            wday: wday as WeekDay,
            items: parsedFoods
          };
        });
    });

  return { name: "Seaside", url, days: _.flatten(menus) };
};
