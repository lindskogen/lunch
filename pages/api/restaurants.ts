import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../../lib/lunch-fetcher/fetch-restaurants";

const restaurants = async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};

export default restaurants;
