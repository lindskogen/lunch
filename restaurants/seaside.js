const _ = require("lodash");
const { isWeekday, isNotNull, isNotZero } = require("../lib/utils");

const url = "http://seaside.kvartersmenyn.se/";
module.exports.url = url;

module.exports.parseHtml = $ => {
  const menus = $(".day .meny")
    .toArray()
    .map(item => {
      const allNodes = item.childNodes
        .filter(i => i.tagName === "strong" || i.nodeValue)
        .map(item => $(item).text());

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const parsedFoods = foods.map(food => {
            const res = /^([A-Ã–]+( [0-9])?) ?:?([^0-9]*)/.exec(food);
            if (res) {
              const [all, title, number, name] = res;
              return { title, name: name.trim() };
            } else {
              return null;
            }
          });

          return {
            wday,
            items: parsedFoods
          };
        });
    });

  return { name: "Seaside", url, days: _.flatten(menus) };
};
