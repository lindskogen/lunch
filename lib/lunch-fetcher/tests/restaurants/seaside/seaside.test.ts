import * as cheerio from "cheerio";
import { parseHtml } from "../../../restaurants/seaside";

import mock1 from "./mock1";

describe("seaside", () => {
  describe("when provided with mock 1", () => {
    const result = parseHtml(cheerio.load(mock1));
    it("should work", () => {
      expect(result).toMatchSnapshot();
    });

    it("should parse 5 days", () => {
      expect(result.days).toHaveLength(5);
    });

    it("should remove colon from title and name", () => {
      result.days.forEach(day => {
        day.items.forEach(item => {
          expect(item.title).not.toMatch(/:/);
          expect(item.name).not.toMatch(/:/);
        });
      });
    });

    it("should have no empty names", () => {
      result.days.forEach(day => {
        day.items.forEach(item => {
          expect(item.name).not.toHaveLength(0);
        });
      });
    });
  });
});
