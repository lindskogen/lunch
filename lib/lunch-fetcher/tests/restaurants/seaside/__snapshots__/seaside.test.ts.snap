// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seaside when provided with mock 1 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Isterband med dillstuvad potatis, picklade rödbetor och hemgjord senap",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Het chiligryta med nötfärs, bönor och paprika samt picklad rödlök, crème fraiche och ris",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Rödbetsgryta med fullkornsbulgur och chèvreklick",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Hamburgare med ost, pommes frites, saltgurka och tillbehör",
          "title": "GRILL",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Stekt fläsk med bruna bönor eller löksås serveras med kokt potatis",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Mandel och fröpanerad sejfilé med curryrostad blomkål och spenatsallad",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Rödbetsgryta med fullkornsbulgur och chèvreklick",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Hamburgare med ost, pommes frites, saltgurka och tillbehör",
          "title": "GRILL",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Panerad torskfilé med skarpsås, dillslungad potatis och citron",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Adobokryddad kyckling med karibisk bönsallad och rostad potatis",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Rödbetsgryta med fullkornsbulgur och chèvreklick",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Hamburgare med ost, pommes frites, saltgurka och tillbehör",
          "title": "GRILL",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Ärtsoppa med fläsk och korv samt pannkakor med sylt och vispad grädde",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Nötfärsbiff med karamelliserad lök, rödvinssås, gelé och stekt potatis",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Rostade grönsaker med libanesisk kryddning, matvete och hummus",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Hamburgare med ost, pommes frites, saltgurka och tillbehör",
          "title": "GRILL",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Helstekt fläskfilé med råstekt potatis, pepparsås och haricots verts",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Panerad rödspätta med skaldjursmajonnäs, citron och dillslungad potatis",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Rostade grönsaker med libanesisk kryddning, matvete och hummus",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Hamburgare med ost, pommes frites, saltgurka och tillbehör",
          "title": "GRILL",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Seaside",
  "url": "http://seaside.kvartersmenyn.se/",
}
`;

exports[`seaside when provided with mock 2 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Strimlad biff och bambuskott i soyasås med rättikasallad serveras med ris",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Gräddstuvad pytt i panna med stekt ägg och inlagda rödbetor",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Marockansk rotfruktsgryta med couscous samt aubergine & valnötsröra (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein & plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv & bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Stekt fläsk med bruna bönor eller löksås samt kokt potatis",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Fiskgryta med bouillabaisesås, rotfrukter och fänkål samt aioli",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Chili sin carne serveras med crème fraiche, picklad rödlök och råris",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Kycklingfilé med medelhavsgrönsaker,  aubergine & bönröra samt örtbulgur",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Panerad rödspätta med curryremouladesås, citron & kokt potatis",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Chili sin carne serveras med crème fraiche, picklad rödlök och råris",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Ärtsoppa med fläsk och köttkorv samt pannkakor med sylt och vispad grädde",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Husets lasagne serveras med  tomat, ruccola och rödlökssallad",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Lins och böngryta med sambal, koriander och kokosmjölk serveras med ris (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Wallenbergare med potatispuré, gröna ärtor, rårörda lingon & brynt smör",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Örtbakad sejfilé med lökfräst bladspenat, dillslungad potatis och smörsås",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Lins och böngryta med sambal, koriander och kokosmjölk serveras med ris (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Seaside",
  "url": "http://seaside.kvartersmenyn.se/",
}
`;
